<a href="https://colab.research.google.com/github/SangMin316/EEG_Data/blob/main/Sleep_defx_to_pickle.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
!pip install pyedflib

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 2.4 MB 8.6 MB/s 


In [ ]:
from pyedflib import highlevel
import numpy as np
import glob

path1 = '/content/drive/MyDrive/sleep_edfx/data/sleep-cassette'
path2 = '/content/drive/MyDrive/sleep_edfx/data/sleep-telemetry'
data_list_cassette = glob.glob(path1+'/**')
data_list_telemetry = glob.glob(path2+'/**') 

print('number of cattette files',len(data_list_cassette))
print('number of telemetry files',len(data_list_telemetry))


#First, investigate cassette files.
trains = [x for x in data_list_cassette if x.endswith('PSG.edf')]
labels = [x for x in data_list_cassette if x.endswith('Hypnogram.edf')]

print('number of train samples:',len(trains))
print('number of labels:',len(labels))

print(highlevel.read_edf(trains[0]))
print('')
print(highlevel.read_edf(labels[0]))

number of cattette files 306
number of telemetry files 88
number of train samples: 153
number of labels: 153
([array([-12.93333333, -12.04444444, -10.88888889, ...,  34.44444444,
        75.95555556,  77.82222222]), array([ 6.43711844,  0.83247863,  5.2021978 , ..., 26.86080586,
       -5.34212454, 22.01611722]), array([ 220.54993895,  220.06471306,  214.72722833, ..., -101.64004884,
       -104.06617827,  -94.84688645]), array([130., 129., 112., ...,  53.,  41.,  45.]), array([3.498, 3.512, 3.51 , ..., 3.05 , 3.024, 3.058]), array([14.1794461 , 14.15335317, 14.12039368, ..., 14.11032273,
       14.13641566, 14.13504234]), array([1037., 1012.,  985., ...,  858.,  880.,  896.])], [{'label': 'EEG Fpz-Cz', 'dimension': 'uV', 'sample_rate': 3000.0, 'sample_frequency': 3000.0, 'physical_max': 182.0, 'physical_min': -182.0, 'digital_max': 2047, 'digital_min': -2048, 'prefilter': 'HP:0.5Hz LP:100Hz [enhanced cassette BW]', 'transducer': 'Ag-AgCl electrodes'}, {'label': 'EEG Pz-Oz', 'dimension

cattette 파일에 153개의 sample이 존재합니다.

highlevel.read_edf return 값이 signals, signal_headers, header 3개입니다.

따로 따로 값을 출력해보겠습니다.



In [ ]:
signals = highlevel.read_edf(trains[0])[0]
#signals : np.ndarray or list
#the signals of the chosen channels contained in the EDF.

print(signals)
print(len(signals))
for i in range(len(signals)):
    print(len(signals[i]))

[array([-12.93333333, -12.04444444, -10.88888889, ...,  34.44444444,
        75.95555556,  77.82222222]), array([ 6.43711844,  0.83247863,  5.2021978 , ..., 26.86080586,
       -5.34212454, 22.01611722]), array([ 220.54993895,  220.06471306,  214.72722833, ..., -101.64004884,
       -104.06617827,  -94.84688645]), array([130., 129., 112., ...,  53.,  41.,  45.]), array([3.498, 3.512, 3.51 , ..., 3.05 , 3.024, 3.058]), array([14.1794461 , 14.15335317, 14.12039368, ..., 14.11032273,
       14.13641566, 14.13504234]), array([1037., 1012.,  985., ...,  858.,  880.,  896.])]
7
8430000
8430000
8430000
84300
84300
84300
84300


signals에는 list 요소가 7개이고, 각 요소에 array(신호)들이 들어있습니다.



In [ ]:
signal_headers = highlevel.read_edf(trains[0])[1]
#signal_headers : list
#one signal header for each channel in the EDF.

print('number of channels:',len(signal_headers))
print('\n')

for i in range(len(signal_headers)):
    print(signal_headers[i])
    print(len(signal_headers[i]))
    print('\n')

number of channels: 7


{'label': 'EEG Fpz-Cz', 'dimension': 'uV', 'sample_rate': 3000.0, 'sample_frequency': 3000.0, 'physical_max': 182.0, 'physical_min': -182.0, 'digital_max': 2047, 'digital_min': -2048, 'prefilter': 'HP:0.5Hz LP:100Hz [enhanced cassette BW]', 'transducer': 'Ag-AgCl electrodes'}
10


{'label': 'EEG Pz-Oz', 'dimension': 'uV', 'sample_rate': 3000.0, 'sample_frequency': 3000.0, 'physical_max': 195.0, 'physical_min': -194.0, 'digital_max': 2047, 'digital_min': -2048, 'prefilter': 'HP:0.5Hz LP:100Hz [enhanced cassette BW]', 'transducer': 'Ag-AgCl electrodes'}
10


{'label': 'EOG horizontal', 'dimension': 'uV', 'sample_rate': 3000.0, 'sample_frequency': 3000.0, 'physical_max': 994.0, 'physical_min': -993.0, 'digital_max': 2047, 'digital_min': -2048, 'prefilter': 'HP:0.5Hz LP:100Hz [enhanced cassette BW]', 'transducer': 'Ag-AgCl electrodes'}
10


{'label': 'Resp oro-nasal', 'dimension': '', 'sample_rate': 30.0, 'sample_frequency': 30.0, 'physical_max': 3000.0, 'physical_m

signal_headers에는 신호에 대한 정보들이 존재합니다.

논문과 웹사이트에서는 sample rate가 100Hz, 1Hz라고 나오는데, 3000, 30으로 나옵니다.

이는 epoch 사이즈가 30초이기에, 30초당 3000번, 즉 100Hz라고 이해하면 될것 같습니다.

In [ ]:
header = highlevel.read_edf(trains[0])[2]
#header : dict
#the main header of the EDF file containing meta information.
print(header)
print(len(header))

{'technician': '', 'recording_additional': '', 'patientname': '', 'patient_additional': '', 'patientcode': '', 'equipment': '', 'admincode': '', 'gender': '', 'startdate': datetime.datetime(1991, 9, 26, 15, 0), 'birthdate': '', 'annotations': []}
11


header에는 datetime 말고는 정보가 없습니다.

trains data에 말고 label_data에 header에 정보가 있습니다.

labels 데이터를 살펴보겠습니다.

In [ ]:
signals = highlevel.read_edf(labels[0])[0]
#signals : np.ndarray or list
#the signals of the chosen channels contained in the EDF.
print(signals)

signal_header = highlevel.read_edf(labels[0])[1]
print(signal_header)

[]
[]


signals와 signal_header에 아무것도 없는 것을 확인 할 수 있습니다.



In [ ]:
header = highlevel.read_edf(labels[0])[2]
print(len(header))
print(header)

11
{'technician': '', 'recording_additional': '', 'patientname': 'Female 26yr', 'patient_additional': '', 'patientcode': '', 'equipment': '', 'admincode': '', 'gender': 'Female', 'startdate': datetime.datetime(1989, 4, 10, 16, 50), 'birthdate': '', 'annotations': [[0.0, 26130.0, 'Sleep stage W'], [26130.0, 30.0, 'Sleep stage 1'], [26160.0, 60.0, 'Sleep stage W'], [26220.0, 240.0, 'Sleep stage 1'], [26460.0, 30.0, 'Sleep stage W'], [26490.0, 90.0, 'Sleep stage 1'], [26580.0, 1050.0, 'Sleep stage 2'], [27630.0, 30.0, 'Sleep stage 3'], [27660.0, 150.0, 'Sleep stage 2'], [27810.0, 390.0, 'Sleep stage 3'], [28200.0, 90.0, 'Sleep stage 2'], [28290.0, 90.0, 'Sleep stage 3'], [28380.0, 30.0, 'Sleep stage 2'], [28410.0, 60.0, 'Sleep stage W'], [28470.0, 30.0, 'Sleep stage 1'], [28500.0, 30.0, 'Sleep stage W'], [28530.0, 30.0, 'Sleep stage 1'], [28560.0, 30.0, 'Sleep stage 2'], [28590.0, 30.0, 'Sleep stage 1'], [28620.0, 690.0, 'Sleep stage 2'], [29310.0, 30.0, 'Sleep stage 3'], [29340.0, 60.0, 



---


labels data에 주석이 달려 있습니다.

[시작 시간, 지속시간, 수면 상태]

이런식으로 labeling되어 있습니다.

label 구성: "Sleep stage ?", "Sleep stage W", "Sleep stage 1", "Sleep stage 2", "Sleep stage 3", "Sleep stage 4", "Sleep stage R"

Get information EDF files by other mathod


---



another method

In [ ]:
import pyedflib
edf = pyedflib.EdfReader(trains[1])
print('labels:'+str(edf.getSignalLabels()))
print("Duaration:"+str(edf.getFileDuration()))
print("Freq.:"+str(edf.getSampleFrequencies()))
print("N-Sample(=Freq x Duaration):"+str(edf.getNSamples()))
print("Date:"+str(edf.getStartdatetime()))

labels:['EEG Fpz-Cz', 'EEG Pz-Oz', 'EOG horizontal', 'Resp oro-nasal', 'EMG submental', 'Temp rectal', 'Event marker']
Duaration:80700
Freq.:[3000 3000 3000   30   30   30   30]
N-Sample(=Freq x Duaration):[8070000 8070000 8070000   80700   80700   80700   80700]
Date:1989-07-12 16:25:00


EEG 신호만 따로 추출해서 파일을 저장해주겠습니다.


In [ ]:
trains1 = [x for x in data_list_telemetry if x.endswith('PSG.edf')]
trains.append(trains1)
print(len(trains))

154


In [ ]:
edf = highlevel.read_edf(trains[0])[0]
a = edf[0]
b = edf[1]
data = np.array([a,b])
print(data.shape)
data

(2, 8430000)


array([[-12.93333333, -12.04444444, -10.88888889, ...,  34.44444444,
         75.95555556,  77.82222222],
       [  6.43711844,   0.83247863,   5.2021978 , ...,  26.86080586,
         -5.34212454,  22.01611722]])

In [ ]:
import pickle

for i in range(len(trains)):
    edf = highlevel.read_edf(trains[i])[0]
    a = edf[0] 
    b = edf[1]
    data = np.array([edf[0],edf[1]]) # # EEG Fpz-Cz signal, EEG Pz-Oz signal 
    a = '/content/drive/MyDrive/sleep_edfx/sleep_edfx_CT+SC/' + '%i.pickle' %i
    with open(file = a, mode='wb') as f:
      pickle.dump(data, f)